python/F3DPythonBindings.cxx

```diff
@@ -102,108 +102,148 @@ PYBIND11_MODULE(pyf3d, module)
+    .def("__setitem__", py::overload_cast<const std::string&, bool>(&f3d::options::set))
+    .def("__setitem__", py::overload_cast<const std::string&, int>(&f3d::options::set))
+    .def("__setitem__", py::overload_cast<const std::string&, double>(&f3d::options::set))
+    .def(
+      "__setitem__", py::overload_cast<const std::string&, const std::string&>(&f3d::options::set))
+    .def("__setitem__",
+      py::overload_cast<const std::string&, const std::vector<int>&>(&f3d::options::set))
+    .def("__setitem__",
+      py::overload_cast<const std::string&, const std::vector<double>&>(&f3d::options::set))
+    .def("__getitem__",
+      [](f3d::options& opts, const std::string key)
+      {
+#define TRY(getter)                                                                                \
+  try                                                                                              \
+  {                                                                                                \
+    auto v = getter(key);                                                                          \
+    return py::cast(v);                                                                            \
+  }                                                                                                \
+  catch (const f3d::options::inexistent_exception&)                                                \
+  {                                                                                                \
+    throw pybind11::key_error(key);                                                                \
+  }                                                                                                \
+  catch (const f3d::options::incompatible_exception&)                                              \
+  {                                                                                                \
+  }
+        TRY(opts.getAsBoolRef)
+        TRY(opts.getAsIntRef)
+        TRY(opts.getAsDoubleRef)
+        TRY(opts.getAsStringRef)
+        TRY(opts.getAsDoubleVectorRef)
+        TRY(opts.getAsIntVectorRef)
+        throw pybind11::key_error(key);
+#undef TRY
+      })
+    .def("__len__", [](f3d::options& opts) { return opts.getNames().size(); })
+    .def(
+      "__iter__",
+      [](f3d::options& opts)
+      {
+        const auto names = py::cast(opts.getNames()); // won't work without cast
+        return make_iterator(names);
+      },
+      py::keep_alive<0, 1>())
+    .def("keys", &f3d::options::getNames) // to do `dict(options)`
+    // TODO: keep the get_as_?
+    //  .def("get_as_bool", &f3d::options::getAsBool)
+    //  .def("get_as_int", &f3d::options::getAsInt)
+    //  .def("get_as_double", &f3d::options::getAsDouble)
+    //  .def("get_as_string", &f3d::options::getAsString)
+    //  .def("get_as_int_vector", &f3d::options::getAsIntVector)
+    //  .def("get_as_double_vector", &f3d::options::getAsDoubleVector)
```

Comment:
todo

Reply:
Note that `__getitem__` (ie. `v = options[k]`) replaces all of these, so technically they're not needed. They _could_ be kept for "compatibility" with the c++ API, whether they _should_ be kept is up for debate :)

Reply:
I vote for omitting them. Retrocompat is ensured by the deprecated functions, but using `__getitem__` should be the preferred way.

Reply:
fine by me, just remove the todo and the dead code.